### PR TITLE
Feature/hydrodynamics

### PR DESCRIPTION
**Pull request by arjo129**
🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

# 🎉 New feature

Closes #<NUMBER>

## Summary
This PR adds support for hydrodynamic simulation of underwater vehicles. It does so by introducing two new plugins. The first consists of a new Hydrodynamics system and a new Thruster system. These systems are based on the equations described in Fossen's "Guidance and Control of Ocean Vehicles".
loop
I've added an example world and tutorial giving more details of how to use the new plugins.

## Test it
An example configuration is provided in the examples folder. The example uses the LiftDrag plugin to apply steering controls. It also uses the thruster plugin to propel the craft and the buoyancy plugin for buoyant force. To run the example run.
```
ign gazebo auv_controls.sdf
```
To control the rudder of the craft run the following
```
ign topic -t /model/tethys/joint/vertical_fins_joint/0/cmd_pos -m ignition.msgs.Double -p 'data: -0.17'
```
To apply a thrust you may run the following command
```
ign topic -t /model/tethys/joint/propeller_joint/cmd_pos -m ignition.msgs.Double -p 'data: -31'
```
The vehicle should move in a circle.

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [x] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

File: src/systems/thruster/Thruster.cc [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+  ignition::gazebo::Link link(this->dataPtr->linkEntity);
+
+  auto pose = worldPose(this->dataPtr->linkEntity, _ecm);
+
+  // TODO(arjo129): add logic for custom coordinate frame
+  auto unitVector = pose.Rot().RotateVector(
+    this->dataPtr->jointAxis.Normalize());
+
+  std::lock_guard<std::mutex> lock(this->dataPtr->mtx);
```
Comment (caguero) [inline src/systems/thruster/Thruster.cc:239]: Do you need to hold the lock until the end of the function? You can probably use a new scope `{` `}` and surround just the line to read `this->dataPtr->thrust`.

**Review by caguero — CHANGES_REQUESTED** [review]
This looks pretty good! Thanks for the tutorial and examples. I added some minor suggestions but it's very close.

**Comment by caguero** [comment]
I took the liberty of adding a few commits (22c0f13 and 057d9e2) to tweak the style and minor typos.

File: src/systems/hydrodynamics/Hydrodynamics.cc [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+}
+
+void AddWorldPose (
+  const ignition::gazebo::Entity &_entity,
+  ignition::gazebo::EntityComponentManager &_ecm)
+{
+  if (!_ecm.Component<ignition::gazebo::components::WorldPose>(
+      _entity))
+  {
+    _ecm.CreateComponent(_entity,
```
[the remark this answers is not in the document]
Comment (arjo129) [reply src/systems/hydrodynamics/Hydrodynamics.cc]: `WorldPose` is used here: https://github.com/ignitionrobotics/ign-gazebo/pull/749/files/d3049970c7f3feac28b5c96e9d79e6ebee4d3f69#diff-1a4f185dacd3135bdd14923737f61e9be4c2c6358fd334e8c8ed5d9452f77450R244-R248 to calculate the transform between local and global frame.

**Review by caguero — APPROVED** [review]

**Comment by Pedro-Roque** [comment]
@arjo129 - opening a discussion on naming for this plugin. I understand that Thrusters are used for underwater propellers enclosed in a cylinder to propel a vehicle forward. However, thrusters are also used in the context of a spacecraft, in which case the dynamics are considerably different. What name should we give to the later case?

**Comment by arjo129** [comment]
Hi @Pedro-Roque thanks for pointing this out. As this has been merged and in use fore several years now it would be difficult to change. I recommend opening a ticket on the issues page so that we can track this conversation. I guess this would have been better served if it was called "maritime-thruster" instead of "thruster".

**Pull request merged**